workload/workload.go

```diff
@@ -0,0 +1,200 @@
+package workload
+
+import (
+	"context"
+	"fmt"
+	"github.com/couchbase/gocb/v2"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.uber.org/zap"
+	"log"
+	"math/rand"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+// A spectroperf Workload is defined by:
+//   - The documents it operates on
+//   - The Operations it performs on those documents
+//   - The probability matrix defining the likelihood of one operation being followed by another
+type Workload interface {
+	// GenerateDocument creates a random document appropriate for the workload
+	GenerateDocument(id string) DocType
+	// Operations returns The list of operations that a workload supports
+	Operations() []string
+	// Probabilities returns the probability matrix for the workload
+	Probabilities() [][]float64
+	// Returns a map of operations to workload functions
+	Functions() map[string]func(ctx context.Context, rctx Runctx) error
```

Comment:
Not a huge fan of having Functions() and Operations(), since the operations are just the keys of the function map. However GO maps have order guarantee so to ensure the probabilities line up with the operations we need the operations stored in an array.